### PR TITLE
dE/dx updates in miniAOD

### PR DESCRIPTION
**Pull request by gpetruc**
Add hit-level dE/dx information to miniAOD for selected tracks:
 * high pT tracks for EXO studies: p<sub>T</sub> > 50 GeV, loose dxy, dz cuts, loose isolation (no isolation for p<sub>T</sub> > 200 GeV)
 * disappearing tracks for SUSY: p<sub>T</sub> > 15 GeV (10 GeV for pixel-only tracks), tight dxy, dz cuts, lepton veto, isolation cut.

Expected number of selected tracks per event on some MC and data files:

PD | EXO | SUSY | TOTAL
-- | -- | -- | --
Data_SingleMuon | 0.18 | 0.06 | 0.23
Data_SingleElectron | 0.23 | 0.09 | 0.31
Data_MET | 0.06 | 0.04 | 0.09
Data_JetHT | 0.21 | 0.10 | 0.28
MC_DYJets | 0.11 | 0.10 | 0.20
MC_TT | 0.27 | 0.15 | 0.41
MC_WJets | 0.04 | 0.04 | 0.08
MC_QCD | 0.18 | 0.06 | 0.21

 Data is run 299649 ( ~ 1.6E34 lumi )

@mmasciov @arizzi @jozzez1 @bjmarsh

### CODE REVIEW

**Comment by slava77** [comment]
@cmsbuild please test

**Comment by perrotta** [comment]
Error is related to this PR (in FastSim)

**Comment by gpetruc** [comment]
fixed. that required however a change to `Configuration/Applications` because the fast sim customization was put directly there.

**Comment by perrotta** [comment]
please test

File: PhysicsTools/PatAlgos/plugins/PATIsolatedTrackProducer.cc [hunk 1]
```diff
@@ -435,10 +539,112 @@ void pat::PATIsolatedTrackProducer::getIsolation(const LorentzVector& p4,
...
+//get overlap of isolated track with a PF lepton
+bool pat::PATIsolatedTrackProducer::getPFLeptonOverlap(const LorentzVector& p4, const pat::PackedCandidateCollection *pc, int pc_idx) const
+{
+        bool isOverlap = false;
+	float dr_min = pflepoverlap_DR_;
+	int id_drmin = 0;
+        for(pat::PackedCandidateCollection::const_iterator pf_it = pc->begin(); pf_it != pc->end(); pf_it++){
+//	    // Count candidate itself (commenting out lines below): exclude isolated track that is a pf lepton
+//            if(int(pf_it - pc->begin()) == pc_idx)
+//                continue;
```
Comment (slava77) [inline PhysicsTools/PatAlgos/plugins/PATIsolatedTrackProducer.cc]: is this commented out code needed?
Please remove or add comments inline in the code why the commented out block is relevant

Comment (slava77) [reply PhysicsTools/PatAlgos/plugins/PATIsolatedTrackProducer.cc]: I guess one can argue that the inline comment is in place

Comment (slava77) [reply PhysicsTools/PatAlgos/plugins/PATIsolatedTrackProducer.cc]: it looks like pc_idx argument and this commented out piece can be removed

Comment (gpetruc) [reply PhysicsTools/PatAlgos/plugins/PATIsolatedTrackProducer.cc]: @mmasciov just to be sure, the intended behaviour is that isolated tracks that are pf candidates that are leptons should always fail `getPFLeptonOverlap`, right? (provided they pass fromPV)

Comment (gpetruc) [reply PhysicsTools/PatAlgos/plugins/PATIsolatedTrackProducer.cc]: @slava77 : done

**Comment by slava77** [comment]
@cmsbuild please test

File: PhysicsTools/PatAlgos/plugins/PATIsolatedTrackProducer.cc (hunk 1 above)
Comment (mmasciov) [reply PhysicsTools/PatAlgos/plugins/PATIsolatedTrackProducer.cc]: @gpetruc : that is correct, this is why I wanted to count the isolated track itself.

**Comment by slava77** [comment]
@cmsbuild please test

there were file IO problems in the failing job

**Comment by slava77** [comment]
+1

for https://github.com/cms-sw/cmssw/pull/20566	b4f0a17
- extend content of pat::IsolatedTrack and also save detailed DeDx info for selected tracks in miniAOD isolatedTracks
- jenkins tests pass and comparisons with baseline show an expected change in the event content and no relevant differences in plots (new objects are not monitored)

Some plots from 3K events ttbar PU35

roughly expected number of overlaps with leptons (\~2K in a 3K event set)
all_orig-pat-from-aodvssign967-pat-from-aod_ttbar13tevpu2017wf10224p0c_patisolatedtracks_isolatedtracks__pat_obj_pflepoverlap

about 0.4 tracks/event with dedxhitInfos
all_orig-pat-from-aodvssign967-pat-from-aod_ttbar13tevpu2017wf10224p0c_recodedxhitinfos_isolatedtracks__pat_obj_size

Total file size increase  is 211 bytes/event
- 7     in  recoDeDxHitInfosedmAssociation_isolatedTracks.
- 186  in recoDeDxHitInfos_isolatedTracks
-  (490.3 ->   508.3 )        18 in    patIsolatedTracks_isolatedTracks

- CPU of the isolatedTracks module is up from 30 ms to 50 ms /event in my test with d56bcbd
    - the final version had some speedup added (deltaR moved to a later place where it's needed) so, my guess is the final is about 45 ms/evt

**Comment by davidlange6** [comment]
merge

**Comment by slava77** [comment]
this PR has broken the tau embedding unit test (testTauEmbeddingProducers) :

```
   [2] Prefetching for module PoolOutputModule/'MINIAODSIMoutput'
   [3] Calling method for module PATIsolatedTrackProducer/'isolatedTracks'
Exception Message:
ValueMap: no associated value for given product and index

```
this can be reproduced by checkout of TauAnalysis/MCEmbeddingTools and then
```scram b runtests```

**Pull request merged**